tests/Context/TagsContextTest.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+/*
+ * This file is part of Raven.
+ *
+ * (c) Sentry Team
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Raven\Tests\Context;
+
+use PHPUnit\Framework\TestCase;
+use Raven\Context\TagsContext;
+
+class TagsContextTest extends TestCase
+{
+    /**
+     * @dataProvider mergeDataProvider
+     */
+    public function testMerge($data, $recursive, $expectedData, $expectedExceptionMessage)
+    {
+        if (null !== $expectedExceptionMessage) {
+            $this->expectException(\InvalidArgumentException::class);
+            $this->expectExceptionMessage($expectedExceptionMessage);
+        }
+
+        $context = new TagsContext(['foo' => 'bar', 'bar' => 'foo']);
+        $context->merge($data, $recursive);
+
+        $this->assertEquals($expectedData, $context->toArray());
+    }
+
+    public function mergeDataProvider()
+    {
+        return [
+            [
+                ['foo' => 'baz', 'baz' => 'foo'],
+                false,
+                ['foo' => 'baz', 'bar' => 'foo', 'baz' => 'foo'],
+                null,
+            ],
+            [
+                ['foo' => 'bar'],
+                true,
+                null,
+                'The tags context does not allow recursive merging of its data.',
+            ],
+            [
+                ['foo' => new \stdClass()],
+                false,
+                null,
+                'The $data argument must contains a simple array of string values.',
```

Comment:
Should we consider also objects that have `__toString` as valid too?

Reply:
I did not consider them to not have to force the cast to `string` like it was done in `1.x` It's not a problem to add the approprite checks to consider objects, but this would break any typehint we could add in the future (this argument is not strictly related to the methods available at the moment)

https://github.com/getsentry/sentry-php/blob/351cb63da6fc40dbb2a8324e0b3fdff2e7dcc55b/lib/Raven/Client.php#L945

Reply:
I would like to avoid the `@` operator too. But a `is_object && method_exists` could be a good alternative, since the `__toString` method cannot throw exceptions.

Reply:
In which cases it makes sense to add a tag which is an object (even if then it will be convered to a string)? Wouldn't it be better to know that tags can be strings and let users do the cast if they want?

Reply:
The easiest example that I can think of is an `User` class with the username as a `__toString`. I would prefer the string too, but we are on PHP 5.6+, so no scalar type available... Hence that's why I would manually consider castable objects.

Reply:
>but we are on PHP 5.6+, so no scalar type available...

But when we will switch to PHP 7+ it will (`string`, at least). I don't really see why an user could not cast the object before adding it as a tag to the context, it seems a bad practice to me to let users pass an object when it's logic to think that a tag is a string